plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -968,19 +968,8 @@ public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, Con
                     .orElseGet(() -> catalog.defaultTableLocation(session, tableMetadata.getTable()));
         }
         transaction = newCreateTableTransaction(catalog, tableMetadata, session, replace, tableLocation);
-        Location location = Location.of(transaction.table().location());
-        TrinoFileSystem fileSystem = fileSystemFactory.create(session.getIdentity(), transaction.table().io().properties());
-        try {
-            if (!replace && fileSystem.listFiles(location).hasNext()) {
-                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, format("" +
-                        "Cannot create a table on a non-empty location: %s, set 'iceberg.unique-table-location=true' in your Iceberg catalog properties " +
-                        "to use unique table locations for every table.", location));
-            }
-            return newWritableTableHandle(tableMetadata.getTable(), transaction.table(), retryMode);
-        }
-        catch (IOException e) {
-            throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Failed checking new table's location: " + location, e);
-        }
+        catalog.assertNewTableLocationIsEmpty(session, replace, transaction.table().location(), fileSystemFactory, transaction.table().io().properties());
```

Comment:
Should this logic just be part of `TrinoCatalog#newCreateTableTransaction()`? IIRC we're basically validating if the Transaction object is "valid" in the sense that the location returned as part of the transaction is ready to be written. So it seems sensible to do this as part of the txn creation itself. The only other place these 2 methods are used is in `MigrateProcedure#doMigrate()` which currently lacks this emptiness check but looks like it should probably have it. Plus the emptiness check is only relevant if `replace == false`, which naturally makes `newCreateTableTransaction()` (rather than `newCreateOrReplaceTransaction()`) a good place to implement it.

Even if we keep it separate, I feel like a more generic method name like `validateCreateTableTransaction()` would be more sensible, to provide catalog implementations more flexibility on what validations they perform.

Reply:
Good point then we do not need to pass `replace` where it is not previously needed; the conditional logic in `IcebergUtil#newCreateTableTransaction` takes care of it. Adding a new method in the catalog interface seems redundant.

I am thinking of extracting the checking logic to a method in `IcebergUtil` so they can be reused by each catalog implementation.

Reply:
I took a look at the usage in `MigrateProcedure`, it might not make sense to check the table location because the migration is using the same location as the original Hive table; it just transforms the table format to Iceberg. Maybe we can have a `newMigrateTableTransaction` API?